source/fab/steps/grab/svn.py

```diff
@@ -3,32 +3,160 @@
+        name = name or f'{self.__class__.__name__} {dst} {revision}'.strip()
+        super().__init__(src, dst, name=name, revision=revision)
+
+    def run(self, artefact_store: Dict, config):
+        if not self.tool_available():
+            raise RuntimeError(f"command line tool not available: '{self.command}'")
+        super().run(artefact_store, config)
+
+    @classmethod
+    def tool_available(cls) -> bool:
+        """Is the command line tool available?"""
+        try:
+            run_command([cls.command, 'help'])
+        except FileNotFoundError:
+            return False
+        return True
+
+    def _cli_revision_parts(self):
+        # return the command line argument to specif the revision, if there is one
+        return ['--revision', str(self.revision)] if self.revision is not None else []
 
-    You can include a branch in the URL, for example:
+    def _is_working_copy(self, dst: Union[str, Path]) -> bool:
+        # is the given path is a working copy?
+        try:
+            run_command([self.command, 'info'], cwd=dst)
+        except RuntimeError:
+            return False
+        return True
 
-        GrabSvn(
-            src='https://code.metoffice.gov.uk/svn/lfric/GPL-utilities/trunk',
-            revision=36615, dst='gpl_utils')
+
+class SvnExport(GrabSvnBase):
+    """
+    Export an FCM repo folder to the project workspace.
 
     """
     def run(self, artefact_store: Dict, config):
         super().run(artefact_store, config)
 
-        if not svn:
-            raise ImportError('svn not installed, unable to continue')
+        run_command([
+            self.command, 'export', '--force',
+            *self._cli_revision_parts(),
+            self.src,
+            str(self._dst)
+        ])
+
+
+class SvnCheckout(GrabSvnBase):
+    """
+    Checkout or update an FCM repo.
+
+    .. note::
+        If the destination is a working copy, it will be updated to the given revision, **ignoring the source url**.
+        As such, the revision should be provided via the argument, not as part of the url.
+
+    """
+    def run(self, artefact_store: Dict, config):
+        super().run(artefact_store, config)
+
+        # new folder?
+        if not self._dst.exists():  # type: ignore
+            run_command([
+                self.command, 'checkout',
+                *self._cli_revision_parts(),
+                self.src, str(self._dst)
+            ])
+
+        else:
+            # working copy?
+            if self._is_working_copy(self._dst):  # type: ignore
+                # update
+                # todo: ensure the existing checkout is from self.src?
+                run_command([self.command, 'update', *self._cli_revision_parts()], cwd=self._dst)  # type: ignore
+            else:
+                # we can't deal with an existing folder that isn't a working copy
+                raise ValueError(f"destination exists but is not an fcm working copy: '{self._dst}'")
+
+
+class SvnMerge(GrabSvnBase):
```

Comment:
Be aware that the major point of divergance between Subversion and FCM is in the merge code. FCM implements its own merge algorithm and doesn't use Subversion's. As such it may not be a good idea to try and support FCM and Subversion with the same method.

Reply:
Hmm, there's a risk the user will try to cross-merge. I'm not sure that risk comes from sharing common code though. I think it's a configuration/user error because they have to specify either svn or fcm. I wonder if we can invoke the given tool to see if it's an svn or fcm working copy. Thinking ahead though, I understand we will _want_ to be able to cross merge.

Reply:
There are a lot of problems involved in this regardless of what we do. My immediate concern is that `fcm merge` not only behaves differently to `svn merge` but it also takes different arguments. This is not a problem at the moment but may become one later. I propose we stick with what you have for the moment but be aware that it may need t change later.